ocaml/xapi/xapi_pool_patch.ml

```diff
@@ -470,279 +121,40 @@ let write_patch_applied_db ~__context ?date ?(applied=true) ~self ~host () =
-  get_patch_to_local ~__context ~self;
-
-  finally
-    (fun () -> run_precheck ~__context ~self ~host)
-    (fun () ->
-       (* This prevents leaking space on the slave if the patch is repeatedly uploaded, prechecked and then destroyed *)
-       if not (Pool_role.is_master ()) then begin
-         let path = Db.Pool_patch.get_filename ~__context ~self in
-         Unixext.unlink_safe path;
-       end
-    )
+  forward ~__context ~self (Client.Pool_update.precheck ~host); ""
 
 let apply ~__context ~self ~host =
-  (* 0th, check we're not on oem *)
-  if on_oem ~__context
-  then raise (Api_errors.Server_error (Api_errors.not_allowed_on_oem_edition, ["patch-apply"]));
-
-  (* 1st, check patch isn't already applied *)
-  let () = match get_patch_applied_to ~__context ~patch:self ~host with
-    | Some (r, applied) ->
-      if applied then raise (Api_errors.Server_error(Api_errors.patch_already_applied, [ Ref.string_of self ]))
-      else patch_apply_in_progress ~__context ~patch:self ~host
-    | None -> () in
-
-  (* 2nd, get the patch from the master (no-op if we're the master) *)
-  get_patch_to_local ~__context ~self;
-
-  let path = Db.Pool_patch.get_filename ~__context ~self in
-  (* 3rd, run prechecks *)
-  let (_: string) = run_precheck ~__context ~self ~host in
-
-  (* 4th, apply the patch *)
-  begin
-    write_patch_applied_db ~__context ~applied:false ~self ~host ();
-    match execute_patch path [ "apply" ] with
-    | Success(output, _) ->
-      debug "executing patch successful";
-      write_patch_applied_db ~__context ~self ~host ();
-      (* 5th, write out patch applied file to hd *)
-      write_patch_applied ~__context ~self;
-      (* CA-27145: to handle rolled-up patches, rescan the patch applied directory *)
-      begin
-        try update_db ~__context
-        with e ->
-          (* should never happen but just in case... *)
-          error "Caught exception rescanning patch applied directory: %s" (ExnHelper.string_of_exn e)
-      end;
-      Create_misc.create_patches_requiring_reboot_info ~__context ~host;
-      Create_misc.create_software_version ~__context;
-      output;
-    | Failure(log, exn) ->
-      debug "error from patch application: %s" log;
-      erase_patch_applied_db ~__context ~self ~host ();
-      let error_string = "Backup files already present" in
-      if List.length (Xstringext.String.find_all error_string log) = 0 then
-        raise (Api_errors.Server_error(Api_errors.patch_apply_failed, [log]))
-      else begin
-        let xml = Xml.parse_string log in
-        match xml with
-        | Element ("error", [("errorcode", "PATCH_PRECHECK_FAILED_UNKNOWN_ERROR")], [Element("info", _, [PCData info])]) ->
-          raise (Api_errors.Server_error(Api_errors.patch_apply_failed_backup_files_exist, [info]))
-        | _ ->
-          raise (Bad_precheck_xml "Could not find element info")
-      end
-  end
+  forward ~__context ~self (Client.Pool_update.apply ~host); ""
```

Comment:
Is anything expecting a meaningful return string from this and the other patch-related functions?